### DISCUSSION TITLE
How do FSMStrategy and Event Isolation mechanism work?

### DISCUSSION THREAD

**Discussion by rumrunner0** [aiogram/aiogram#1486] [Q&A]
Hi there!

I'd like to understand how `FSMStrategy` works alongside Event Isolation. My goal is to process events **sequentially** for each user per chat (or for the chat if it's a private chat with the bot) **AND concurrently** for multiple users.

The default settings include `storage=MemoryStorage()` and `fsm_strategy=FSMStrategy.USER_IN_CHAT`. However, these seem to have no effect without `events_isolation=SimpleEventIsolation()`, which is the sole implementation of `BaseEventIsolation`.

On the other hand, when using the dispatcher with the settings below, it appears that updates are handled globally and sequentially — messages from all users are processed one at a time, despite the `FSMStrategy.USER_IN_CHAT` setting.

**Setup**
```python
dispatcher = Dispatcher(
    storage=MemoryStorage(), 
    fsm_strategy=FSMStrategy.USER_IN_CHAT, 
    events_isolation=SimpleEventIsolation()
)
```

**Flow**

**USER1**
_user1\_message1_
_user1\_message2_
_user1\_message3_
**USER2**
_user2\_message1_
_user2\_message2_
_user2\_message3_

This flow is logically incorrect, given the `fsm_strategy=FSMStrategy.USER_IN_CHAT`.

I'd be grateful for any help! Maybe this is a bug or everything is simplier that it looks.

**Comment by mafusailovich** [comment]
Hi. Could you solve this problem? I've recently started learn aiogram and can't understand this, but I found a little difference. My telegram bot receiving photos. It should receive some photo (one or more) and then do something. If I use SimpleEventIsolate() my bot receive many photos in media group like one message (one event), but without this isolation each photo (even if user use media group) handled like one message. I don't understand how it works, but I hope that my experience can help you. If you found a solution send me a link for this.

**Comment by franklin-lol** [comment]
SimpleEventIsolation uses a single global asyncio lock, not a per-user lock. This means it serializes all incoming updates globally — regardless of FSMStrategy. The strategy only controls how FSM storage keys are scoped (per user, per chat, etc.), but it has no effect on concurrency.
To achieve per-user sequential processing with cross-user concurrency, you need a per-key lock instead of a global one:
pythonimport asyncio
from typing import Any, Awaitable, Callable
from aiogram import BaseMiddleware
from aiogram.types import Update

class PerUserIsolation(BaseMiddleware):
    def __init__(self):
        self._locks: dict[int, asyncio.Lock] = {}

    async def __call__(
        self,
        handler: Callable[[Update, dict[str, Any]], Awaitable[Any]],
        event: Update,
        data: dict[str, Any],
    ) -> Any:
        user = data.get("event_from_user")
        if user is None:
            return await handler(event, data)
        
        if user.id not in self._locks:
            self._locks[user.id] = asyncio.Lock()
        
        async with self._locks[user.id]:
            return await handler(event, data)
Register it as outer middleware:
pythondp.update.outer_middleware(PerUserIsolation())
This gives you exactly what you want: events from the same user are processed sequentially, but different users run concurrently.
Note: for long-running bots, consider periodically cleaning up _locks to avoid memory growth.
